### PR TITLE
Only keep validators with an assigned index in `staking_ethereum.flows`

### PR DESCRIPTION
**Pull request by hildobby**
…_ethereum.flows

## Thank you for contributing to Spellbook 🪄
Please open the PR in **draft** and mark as ready when you want to request a review. 

### Description:

[...]

---
quick links for more information:
- [README.md](https://github.com/duneanalytics/spellbook/blob/main/README.md)
- [spellbook docs](https://github.com/duneanalytics/spellbook/tree/main/docs)
- [CONTRIBUTING.md](https://github.com/duneanalytics/spellbook/blob/main/CONTRIBUTING.md)

### CODE REVIEW

**Comment by 0xRobin** [comment]
@hildobby can you give more details on the external dataset you're using here?

**Comment by hildobby** [comment]
> @hildobby can you give more details on the external dataset you're using here?

This dataset only contains validator_index and their corresponding pubkey values. This is because the BLS key check happens offchain due to it being quite compute heavy. So this dataset was fetched from beaconcha.in and using it as an inner join ensures that validator_index is correct while excluding non-valid deposits (there are a few, currently throwing off the validator_index values the current spell calculates by counting incrementally).

TLDR; this fixes a bug I've had for a while and makes the data accurate

**Comment by 0xRobin** [comment]
Ok, thanks for explaining, how is the dataset maintained?

**Comment by hildobby** [comment]
I defined the source (+ included it in a sources.yml) like on here: https://github.com/duneanalytics/spellbook/blob/588ca2762f9436ecf79a57b147f01c8d49ad91c3/dbt_subprojects/tokens/models/prices/prices_solana_day.sql#L25


But it's still failing, is there something I'm missing?

File: dbt_subprojects/hourly_spellbook/models/_sector/staking/ethereum/staking_ethereum_flows.sql [hunk 1]
```diff
@@ -113,3 +99,102 @@ SELECT w.block_time
...
+
+{{ config(
+    
+    schema = 'staking_ethereum',
+    alias = 'flows',
+    post_hook='{{ expose_spells(\'["ethereum"]\',
+                                "sector",
+                                "staking",
+                                \'["hildobby", "0xBoxer"]\') }}')
+}}
```
Comment (0xRobin) [inline dbt_subprojects/hourly_spellbook/models/_sector/staking/ethereum/staking_ethereum_flows.sql:122]: @hildobby you duplicated the full model below, which uses incorrect syntax for the source (hence the compile error)

**Comment by jeff-dude** [comment]
minor note: we are also in sync in TG group for design patterns related to the source in this PR

**Comment by 0xRobin** [comment]
@hildobby we have new data available from Lido which should be able to resolve this bug without us depending on external scripts.
Can you look at the `beacon_lido.validators` table to resolve the index <> pubkey matching?

**Comment by hildobby** [comment]
how should I reference an external dataset in spellbook? where should i place the schema file? and should the jinja part be a ref or a source? @0xRobin @jeff-dude

**Comment by 0xRobin** [comment]
@hildobby from spellbooks view, this is just a normal source like a raw blockchain table.
So you can add a sources yaml file with it and use `{{source(..)}}`

I would add it under `/sources/lido/beacon`

**Comment by hildobby** [comment]
> @hildobby from spellbooks view, this is just a normal source like a raw blockchain table. So you can add a sources yaml file with it and use `{{source(..)}}`
> 
> I would add it under `/sources/lido/beacon`
Screenshot 2025-01-08 at 14 17 10


seems like there's something weird here, what am I missing that's causing the CI error @0xRobin ?

File: sources/_subprojects_outputs/staking/_sources.yml
Comment (jeff-dude) [inline sources/_subprojects_outputs/staking/_sources.yml:1]: you should be good to remove this file now

**Review by jeff-dude — CHANGES_REQUESTED** [review]
looks good, just one minor request to remove old source file

File: sources/_subprojects_outputs/staking/_sources.yml
Comment (hildobby) [reply sources/_subprojects_outputs/staking/_sources.yml:1]: done

**Review by jeff-dude — APPROVED** [review]
